src/mlir/dialect/VeronaOps.cc

```diff
@@ -4,7 +4,114 @@
 #include "VeronaOps.h"
 
 #include "VeronaDialect.h"
+#include "mlir/IR/Builders.h"
 #include "mlir/IR/OpImplementation.h"
+#include "mlir/IR/StandardTypes.h"
+
+#include "llvm/ADT/StringSet.h"
+
+using namespace mlir;
+
+static ParseResult parseClassOp(OpAsmParser& parser, OperationState& state)
+{
+  Region* body = state.addRegion();
+
+  StringAttr nameAttr;
+  if (parser.parseSymbolName(
+        nameAttr, SymbolTable::getSymbolAttrName(), state.attributes))
+    return failure();
+
+  if (parser.parseOptionalAttrDictWithKeyword(state.attributes))
+    return failure();
+
+  if (parser.parseRegion(*body, /*arguments=*/{}, /*argTypes=*/{}))
+    return failure();
+
+  verona::ClassOp::ensureTerminator(*body, parser.getBuilder(), state.location);
+
+  return success();
+}
+
+static void print(OpAsmPrinter& printer, verona::ClassOp op)
+{
+  printer << verona::ClassOp::getOperationName() << ' ';
+  printer.printSymbolName(op.sym_name());
+  printer.printOptionalAttrDict(
+    op.getAttrs(), /*elidedAttrs=*/{SymbolTable::getSymbolAttrName()});
+  printer.printRegion(
+    op.body(), /*printEntryBlockArgs=*/false, /*printBlockTerminators=*/false);
+}
+
+/**
+ * AllocateRegionOp and AllocateObjectOp share similar features, which we verify
+ * here:
+ * - The class referenced by the operation must exist.
+ * - There must be as many field names as there are operands.
+ * - The set of field names must match those specified in the class definition (TODO: not implemented yet)
+ *
+ * TODO: Rather than a templated function, should this be a trait?
+ */
+template<typename Op>
+static LogicalResult verifyAllocationOp(Op op)
+{
+  auto className = op.class_();
+  auto classOp = dyn_cast_or_null<verona::ClassOp>(
+    SymbolTable::lookupNearestSymbolFrom(op.getParentOp(), className));
+  if (!classOp)
+  {
+    return op.emitOpError("class '")
+      << className << "' not found in nearest symbol table";
+  }
+
+  if (op.field_names().size() != op.fields().size())
+  {
+    return op.emitError("The number of operands (")
+      << op.fields().size() << ") for '" << op.getOperationName()
+      << "' op does not match the number of field names ("
+      << op.field_names().size() << ")";
+  }
+
+  return success();
+}
+
+static LogicalResult verify(verona::AllocateRegionOp op)
+{
+  return verifyAllocationOp(op);
+}
+
+static LogicalResult verify(verona::AllocateObjectOp op)
+{
+  return verifyAllocationOp(op);
+}
+
+static LogicalResult verify(verona::ClassOp classOp)
+{
+  Block* body = classOp.getBody();
+  if (body->getNumArguments() != 0)
+    return classOp.emitOpError("expected body to have no arguments");
+
+  // Verify that classes contain only fields, and that there are no duplicates.
```

Comment:
Where do the class methods go?

Reply:
There will eventually be a `verona.method` operation that will be part of classes. I have not included them yet because I want to figure out a few things first that methods need (generics, where-clauses, ...).

Reply:
Right, I'd leave a TODO here, then.

Also, I don't know how MLIR handles functions inside regions. An alternative to functions inside classes would be name mangling of functions outside the classes.

Reply:
Will do.

As far as I understand there is no “MLIR functions”, only Standard/core dialect functions. We can make a `verona.method` operation and give it whatever semantics we want. At some point I suspect we’ll lower everything to top level `func` with mangled names.

Reply:
Right, exactly. MLIR core is basically just the idea of SSA values, types, instructions and regions. The implementation of those are all in dialects, including the "standard" one. The benefit in using the standard ones (including loop, affine, etc) is twofold: first, the conversion to LLVM dialect, then LLVM IR, is done already; second, there are existing passes on those dialects that simplify and optimise, even if they have unknown nodes within.

I see the following lowering stages, long term:
1. AST -> Verona dialect with opaque types
2. Type check and inference -> Verona dialect with Verona types
3. Verona specific optimisations that use the dialect to infer high level semantics
4. Sequence of partial lowering to other dialects, dialect specific optimisations (affine, poly, loop)
5. Full lowering to LLVM dialect, then straight to LLVM IR
6. A select pipeline of LLVM passes, for different optimisation levels (perf, space, debug)
7. Code generation (MIR, ASM, OBJ, link, EXE)

We could have a Verona specific "method" construct, but we should always keep in mind that we will *have* to convert it down to an actual MLIR function with actual arguments and return values. The mangling could very well be done at that stage.

I don't have a strong opinion if this is better to be done at the AST level or MLIR level and I think assuming MLIR is a safer bet with the information we have right now.